### PR TITLE
Add JPQD examples of audio classification and image classification

### PR DESCRIPTION
**Pull request by yujiepan-work**
This PR adds examples of OpenVINO OVTrainer:

(1) QAT and JPQD examples in audio-classification.
(2) JPQD example in image-classification.
(3) renames "teacher_model_or_path" to "teacher_model_name_or_path" in the example scripts. Should be more appropriate

### CODE REVIEW

**Comment by yujiepan-work** [comment]
@AlexKoff88  Hi Alexander, this PR about overflow fix on no-vnni-support machines, is ready for review. @vuiseng9

File: examples/openvino/audio-classification/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Audio classification examples
+
+This folder contains [`run_audio_classification.py`](https://github.com/huggingface/optimum/blob/main/examples/openvino/audio-classification/run_audio_classification.py), a script to fine-tune a 🤗 Transformers model on the 🗣️ [Keyword Spotting subset](https://huggingface.co/datasets/superb#ks) of the SUPERB dataset while applying quantization aware training (QAT). QAT can be easily applied by replacing the Transformers [`Trainer`](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#trainer) with the Optimum [`OVTrainer`]. Any model from our [hub](https://huggingface.co/models) can be fine-tuned and quantized, as long as the model is supported by the [`AutoModelForAudioClassification`](https://huggingface.co/docs/transformers/main/en/model_doc/auto#transformers.AutoModelForAudioClassification) API.
```
Comment (AlexKoff88) [inline examples/openvino/audio-classification/README.md:19]: >>Any model from our [hub](https://huggingface.co/models) can be fine-tuned and quantized, as long as the model is supported by the [`AutoModelForAudioClassification`]
Is this the correct statement really?

Comment (AlexKoff88) [reply examples/openvino/audio-classification/README.md:19]: `quantization aware training` -> `Quantization-Aware Training`

File: examples/openvino/audio-classification/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+### Fintuning Wav2Vec2 on Keyword Spotting with QAT
+
+The following command shows how to fine-tune [wav2vec2-base](https://huggingface.co/facebook/wav2vec2-base) on the 🗣️ [Keyword Spotting subset](https://huggingface.co/datasets/superb#ks) of the SUPERB dataset with Quantization Aware Training (QAT).
```
Comment (AlexKoff88) [inline examples/openvino/audio-classification/README.md:23]: Please make `Quantization Aware Training`aligned.

File: examples/openvino/audio-classification/README.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    --metric_for_best_model accuracy \
+    --save_total_limit 3 \
+    --seed 42
+```
+
+On a single V100 GPU, this script should run in ~45 minutes and yield quantized model with accuracy of **97.5%**.
+
+### Joint Pruning, Quantization and Distillation (JPQD) of Wav2Vec2 on Keyword Spotting
+
+`OVTrainer` also provides advanced optimization workflow via NNCF to structurally prune, quantize and distillation. Following is an example to jointly prune, quantize, distill a Wav2Vec2 model for keyword spotting task. Do take note of additional NNCF config `--nncf_compression_config`. More on how to configure the pruning algorithm, see NNCF documentation [here](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/experimental/torch/sparsity/movement/MovementSparsity.md).
```
Comment (AlexKoff88) [inline examples/openvino/audio-classification/README.md:60]: `to structurally prune, quantize and distillation` ->`to structurally prune, quantize and distill`

Comment (AlexKoff88) [reply examples/openvino/audio-classification/README.md:60]: `More on how to configure the pruning algorithm` -> `For more details on how to configure...`

Comment (AlexKoff88) [reply examples/openvino/audio-classification/README.md:60]: `Do take note of additional NNCF config "--nncf_compression_config"` -> `To enable JPQD optimization use an alternative configuration specified with "--nncf_compression_config"`

File: examples/openvino/image-classification/README.md [hunk 4]
```diff
@@ -39,4 +39,46 @@ python run_image_classification.py \
     --save_total_limit 3 \
     --seed 1337 \
     --output_dir /tmp/beans_outputs/
-```
\ No newline at end of file
+```
+
+### Joint Pruning, Quantization and Distillation on Swin for food101
```
Comment (AlexKoff88) [inline examples/openvino/image-classification/README.md:44]: Please apply the same documentation comments as for README for the audio case.

File: examples/openvino/audio-classification/README.md (hunk 1 above)
Comment (yujiepan-work) [reply examples/openvino/audio-classification/README.md]: > > > Any model from our [hub](https://huggingface.co/models) can be fine-tuned and quantized, as long as the model is supported by the [`AutoModelForAudioClassification`]
> > > Is this the correct statement really?

My answer is yes. `AutoModelForAudioClassification` API is from `transformers` repo, and always yields a normal PyTorch model. When it is fed into `OVTrainer`, it will be wrapped by NNCF to do quantization and finetuning. We follow the statement from existing openvino examples, e.g., in text-classification.

File: examples/openvino/image-classification/README.md (hunk 4 above)
Comment (yujiepan-work) [reply examples/openvino/image-classification/README.md]: We have added the time & accuracy results for image-classification QAT.

https://github.com/huggingface/optimum-intel/blob/e23af292bf1b3643489ef4f6d7ad9c93ec7e370b/examples/openvino/image-classification/README.md?plain=1#L44

**Comment by yujiepan-work** [comment]
@helena-intel Hi, Helena, this is the PR of another two examples for JPQD. Could you help to have a look at it?

File: examples/openvino/audio-classification/README.md (hunk 3 above)
Comment (yujiepan-work) [reply examples/openvino/audio-classification/README.md]: https://github.com/huggingface/optimum-intel/blob/5e63e407b906c8e0bd773ee7d1ee6023f4e52f93/examples/openvino/audio-classification/README.md?plain=1#L60

File: examples/openvino/audio-classification/README.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+On a single V100 GPU, this script should run in ~45 minutes and yield a quantized model with accuracy of **97.5%**.
+
+### Joint Pruning, Quantization and Distillation (JPQD) of Wav2Vec2 on Keyword Spotting
+
+`OVTrainer` also provides advanced optimization workflow via NNCF to structurally prune, quantize and distill. Following is an example of joint pruning, quantization and distillation on Wav2Vec2-base model for keyword spotting task. To enable JPQD optimization, use an alternative configuration specified with `--nncf_compression_config`. For more details on how to configure the pruning algorithm, see NNCF documentation [here](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/experimental/torch/sparsity/movement/MovementSparsity.md).
+
+To run the JPQD example, please install optimum-intel from source. This command will install or upgrade optimum-intel and all necessary dependencies:
+
+```bash
+python -m pip install --upgrade "git+https://github.com/huggingface/optimum-intel.git#egg=optimum-intel[openvino,nncf]"
```
Comment (helena-intel) [inline examples/openvino/audio-classification/README.md:65]:
```suggestion
```
Installing from source should no longer be necessary, JPQD is merged into optimum-intel 1.7.0 and we already have openvino 2023.0 prerelease as requirement too.

File: examples/openvino/image-classification/README.md [hunk 6]
```diff
@@ -39,4 +39,48 @@ python run_image_classification.py \
...
+
+### Joint Pruning, Quantization and Distillation (JPQD) of Swin on food101
+
+`OVTrainer` also provides advanced optimization workflow via NNCF to structurally prune, quantize and distill. Following is an example of joint pruning, quantization and distillation on Swin-base model for food101 dataset. To enable JPQD optimization, use an alternative configuration specified with `--nncf_compression_config`. For more details on how to configure the pruning algorithm, see NNCF documentation [here](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/experimental/torch/sparsity/movement/MovementSparsity.md).
+
+To run the JPQD example, please install optimum-intel from source. This command will install or upgrade optimum-intel and all necessary dependencies:
+
+```bash
+python -m pip install --upgrade "git+https://github.com/huggingface/optimum-intel.git#egg=optimum-intel[openvino,nncf]"
+```
```
Comment (helena-intel) [inline examples/openvino/image-classification/README.md:54]:
```suggestion
```
Same as for audio-classification, should no longer be necessary.

**Comment by helena-intel** [comment]
> @helena-intel Hi, Helena, this is the PR of another two examples for JPQD. Could you help to have a look at it?

Thank you! I will try the examples today.

**Comment by helena-intel** [comment]
> My answer is yes. `AutoModelForAudioClassification` API is from `transformers` repo, and always yields a normal PyTorch model. When it is fed into `OVTrainer`, it will be wrapped by NNCF to do quantization and finetuning. We follow the statement from existing openvino examples, e.g., in text-classification.

That's awesome. The configs we provide are wav2vec2 specific though. Do you know if it would be easy to make them more generic, or to provide configs for other audio classification models?

**Comment by vuiseng9** [comment]
> That's awesome. The configs we provide are wav2vec2 specific though. Do you know if it would be easy to make them more generic, or to provide configs for other audio classification models?

@helena-intel, the current design requires strategy registration in NNCF. Are there any other transformer architecture for audio? could you pls point us? we are in the midst of adding more architecture to NNCF. Thanks!

File: examples/openvino/audio-classification/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+datasets>=1.14.0
```
Comment (AlexKoff88) [inline examples/openvino/audio-classification/requirements.txt:1]: Do we need NNCF installed from source here?

File: examples/openvino/audio-classification/README.md (hunk 5 above)
Comment (yujiepan-work) [reply examples/openvino/audio-classification/README.md]: Thanks for pointing out, I have deleted this part.

File: examples/openvino/image-classification/README.md (hunk 6 above)
Comment (yujiepan-work) [reply examples/openvino/image-classification/README.md]: Deleted. 
https://github.com/huggingface/optimum-intel/blob/bdb501859a81eed71733910b9aeedebe034b07dd/examples/openvino/image-classification/README.md
https://github.com/huggingface/optimum-intel/blob/bdb501859a81eed71733910b9aeedebe034b07dd/examples/openvino/audio-classification/README.md

File: examples/openvino/audio-classification/requirements.txt (hunk 7 above)
Comment (yujiepan-work) [reply examples/openvino/audio-classification/requirements.txt:1]: > Do we need NNCF installed from source here?

Not needed.  JPQD is supported in NNCF 2.4 release, which is already in `setup.py` requirement.

File: examples/openvino/audio-classification/README.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+### Fintuning Wav2Vec2 on Keyword Spotting with QAT
+
+The following command shows how to fine-tune [Wav2Vec2-base](https://huggingface.co/facebook/wav2vec2-base) on the 🗣️ [Keyword Spotting subset](https://huggingface.co/datasets/superb#ks) of the SUPERB dataset with Quantization-Aware Training (QAT). The `OVTrainer` uses a default quantization configuration which should work in many cases, but we can also customize the algorithm details. Here, we quantize the Wav2Vec2-base model with a custom configuration file specified by `--nncf_compression_config`. For more details on the quantization configuration, see NNCF documentation [here](https://github.com/openvinotoolkit/nncf/blob/develop/docs/compression_algorithms/Quantization.md).
```
Comment (yujiepan-work) [inline examples/openvino/audio-classification/README.md:23]: I want to mention a latest change in the PR:

QAT on wav2vec2 model in audio-classification is slightly different from other examples, because wav2vec2 needs a customized quantization config.
The default one in optimum-intel does not fit and will get errors. Therefore, here we mention that users can configure the details of quantization via nncf_compression_config.

**Comment by helena-intel** [comment]
> @helena-intel, the current design requires strategy registration in NNCF. Are there any other transformer architecture for audio? could you pls point us? we are in the midst of adding more architecture to NNCF. Thanks!

For example AST (https://huggingface.co/MIT/ast-finetuned-audioset-10-10-0.4593) and HuBERT (https://huggingface.co/superb/hubert-base-superb-ks).

**Comment by helena-intel** [comment]
@vuiseng9 @yujiepan-work I got a dynamic shapes error when loading the JPQD swin model with the OpenVINO 2023.0 prerelease. It would be good to mention in the README that this model requires static shapes, and give an example of how to load the model (and then might as well add inference too):

```python
from transformers import pipeline, AutoFeatureExtractor
from optimum.intel.openvino import OVModelForImageClassification

feature_extractor = AutoFeatureExtractor.from_pretrained(save_dir)

ov_model = OVModelForImageClassification.from_pretrained(save_dir, compile=False)
ov_model.reshape(1,3,224,224)
ov_model.compile()

ov_pipe = pipeline("image-classification", model=ov_model, feature_extractor=feature_extractor)

outputs = ov_pipe("https://datasets-server.huggingface.co/assets/food101/--/default/validation/98/image/image.jpg")
print(outputs)
```

**Comment by yujiepan-work** [comment]
> @vuiseng9 @yujiepan-work I got a dynamic shapes error when loading the JPQD swin model with the OpenVINO 2023.0 prerelease. It would be good to mention in the README that this model requires static shapes, and give an example of how to load the model (and then might as well add inference too):

Thanks very much for pointing out this error! We are trying to find out the reason and will share more information with you later

**Comment by vuiseng9** [comment]
@helena-intel, thanks for reporting the issue. I think we have two issues right now.
1. The Swin IR is not pruned. can you check if your openvino_model.bin is about 300mb? if yes, we are seeing the same.
2. dynamic shaped IR cannot be compiled.
We suspect that Swin Pytorch definition has hardcoded dimension and caused this issue. We're troubleshooting now,

**Comment by helena-intel** [comment]
> The Swin IR is not pruned. can you check if your openvino_model.bin is about 300mb? if yes, we are seeing the same.

Yes, my model is 352 MB.

> We suspect that Swin Pytorch definition has hardcoded dimension and caused this issue. We're troubleshooting now,

That seems likely. IMO for this kind of model, it's not a critical issue if it doesn't work with dynamic shapes. If you do inference with pipeline the images are resized to a specific shape too, so then you might as well reshape your model to that static shape to not have dynamic shapes overhead. 

@echarlaix if these models have hardcoded dimensions in PyTorch, would it be possible to make OVModelForImageClassification so that on loading the model, we reshape it to the size defined in the feature extractor? It could be an option, defaulting to True.

File: examples/openvino/audio-classification/README.md (hunk 1 above)
Comment (echarlaix) [reply examples/openvino/audio-classification/README.md]: > > > > Any model from our [hub](https://huggingface.co/models) can be fine-tuned and quantized, as long as the model is supported by the [`AutoModelForAudioClassification`]
> > > > Is this the correct statement really?
> 
> My answer is yes. `AutoModelForAudioClassification` API is from `transformers` repo, and always yields a normal PyTorch model. When it is fed into `OVTrainer`, it will be wrapped by NNCF to do quantization and finetuning. We follow the statement from existing openvino examples, e.g., in text-classification.

[Here](https://github.com/huggingface/optimum/blob/v1.7.1/tests/onnxruntime/test_modeling.py#L2368-#L2381) is the list of the architectures that should be supported (architectures for which the ONNX export is validated). The exact list of the architectures supported by `optimum.exporters` can be found [here](https://github.com/huggingface/optimum/blob/1.7.1/optimum/exporters/tasks.py)

**Comment by echarlaix** [comment]
Also the doc build was fixed in https://github.com/huggingface/optimum-intel/pull/228 !

**Comment by AlexKoff88** [comment]
@echarlaix, @helena-intel, what are the substantial comments, in your opinion, that prevent us from merging this?

**Review by AlexKoff88 — APPROVED** [review]

**Comment by echarlaix** [comment]
> Shall we merge? 🎉

Just waiting for @helena-intel's approval
